mava/systems/q_learning/rec_iql.py

```diff
@@ -0,0 +1,662 @@
+# limitations under the License.
+
+import copy
+import time
+from pprint import pprint
+from typing import Any, Callable, Dict, Tuple  # noqa
+
+import chex
+import flashbax as fbx
+import hydra
+import jax
+import jax.lax as lax
+import jax.numpy as jnp
+import optax
+from chex import PRNGKey
+from colorama import Fore, Style
+from flashbax.buffers.flat_buffer import TrajectoryBuffer
+from flax.core.scope import FrozenVariableDict
+from jax import Array
+from jumanji.env import Environment
+from omegaconf import DictConfig, OmegaConf
+
+from mava.evaluator import make_eval_fns
+from mava.networks import RecQNetwork, ScannedRNN
+from mava.systems.q_learning.types import (
+    ActionSelectionState,
+    DDQNParams,
+    InteractionState,
+    LearnerState,
+    Metrics,
+    TrainState,
+    Transition,
+)
+from mava.types import Observation
+from mava.utils import make_env as environments
+from mava.utils.checkpointing import Checkpointer
+from mava.utils.jax import unreplicate_batch_dim, unreplicate_n_dims
+from mava.utils.logger import LogEvent, MavaLogger
+from mava.wrappers import episode_metrics
+
+
+def init(
+    cfg: DictConfig,
+) -> Tuple[
+    Tuple[Environment, Environment],
+    RecQNetwork,
+    optax.GradientTransformation,
+    TrajectoryBuffer,
+    LearnerState,
+    MavaLogger,
+    PRNGKey,
+]:
+    """Initialize system by creating the envs, networks etc.
+
+    Args:
+        cfg: System configuration.
+
+    Returns:
+        Tuple containing:
+            Tuple[Environment, Environment]: The environment and evaluation environment.
+            RecQNetwork: Recurrent Q network
+            optax.GradientTransformation: Optimiser for RecQNetwork.
+            TrajectoryBuffer: The replay buffer.
+            LearnerState: The initial learner state.
+            MavaLogger: The logger.
+            PRNGKey: The random key.
+    """
+    logger = MavaLogger(cfg)
+
+    key = jax.random.PRNGKey(cfg.system.seed)
+    devices = jax.devices()
+
+    def replicate(x: Any) -> Any:
+        """First replicate the update batch dim then put on devices."""
+        x = jax.tree_map(lambda y: jnp.broadcast_to(y, (cfg.system.update_batch_size, *y.shape)), x)
+        return jax.device_put_replicated(x, devices)
+
+    env, eval_env = environments.make(cfg)
+
+    action_dim = env.action_dim
+    num_agents = env.num_agents
+
+    key, q_key = jax.random.split(key, 2)
+    # Shape legend:
+    # T: Time
+    # B: Batch
+    # A: Agent
+    # Make dummy inputs to init recurrent Q network -> need shape (T, B, A, ...)
+    init_obs = env.observation_spec().generate_value()  # (A, ...)
+    init_obs_batched = jax.tree_util.tree_map(
+        lambda x: jnp.repeat(x[jnp.newaxis, ...], cfg.arch.num_envs, axis=0),
+        init_obs,
+    )  # (B, A, ...)
+    init_obs_batched = jax.tree_util.tree_map(
+        lambda x: x[jnp.newaxis, ...], init_obs_batched
+    )  # add time dim (1, B, A, ...)
```

Comment:
We shouldn't need to give the exact batch size as long as the dim exists it should be fine
```suggestion
    init_obs_batched = jax.tree_util.tree_map(lambda x: x[jnp.newaxis, jnp.newaxis, ...], init_obs)  # (B, T, A, ...)
```